tests/integration/v3_watch_restore_test.go

```diff
@@ -154,3 +185,89 @@ func TestV3WatchRestoreSnapshotUnsync(t *testing.T) {
 		}
 	}
 }
+
+type lineCountExpecter struct {
+	doneOnce sync.Once
+	doneCh   chan struct{}
+
+	content string
+	count   int64
+	seen    int64
+}
+
+func newLineCountExpecter(expectedContent string, expectedCount int64) *lineCountExpecter {
+	return &lineCountExpecter{
+		doneCh:  make(chan struct{}),
+		content: expectedContent,
+		count:   expectedCount,
+	}
+}
+
+func (le *lineCountExpecter) Notify(log string) {
+	if !strings.Contains(log, le.content) {
+		return
+	}
+
+	if atomic.AddInt64(&le.seen, 1) >= le.count {
+		le.doneOnce.Do(func() {
+			close(le.doneCh)
+		})
+	}
+}
+
+func (le *lineCountExpecter) wait(timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	select {
+	case <-le.doneCh:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+type testingLogfSubscriber interface {
+	Notify(log string)
+}
+
+// testingLogfMonitor is to monitor t.Logf output.
+type testingLogfMonitor struct {
+	testutil.TB
+
+	mu          sync.RWMutex
+	subscribers map[string]testingLogfSubscriber
+}
+
+func newTestingLogfMonitor(tb testutil.TB) *testingLogfMonitor {
+	return &testingLogfMonitor{
+		TB:          tb,
+		subscribers: make(map[string]testingLogfSubscriber),
+	}
+}
+
+func (m *testingLogfMonitor) addSubscriber(id string, sub testingLogfSubscriber) {
```

Comment:
minor comment: change `addSubscriber` to `register`, and `delSubscriber` to `deregister`